refactor(core): expose host directives to their host through DI

crisbeto opened the pull request:
Exposes the host directives to the host and its descendants through DI. This can be useful, because it allows the host to further configure the host directives.

pkozlowski-opensource commented on packages/core/src/render3/features/host_directives_feature.ts:
```diff
@@ -66,6 +67,10 @@ function findHostDirectiveDefs(
 
+      // Allows for the directive to be injected by the host.
+      diPublicInInjector(
+          getOrCreateNodeInjectorForNode(tNode, lView), tView, hostDirectiveDef.type);
```
What would be great to do in a follow-up PR is to pull the DI logic _out_ of the find functions. It really doesn't seem to belong in here. I know that this is a pre-existing pattern hence I don't think we should block this PR on it. But this is good cleanup to make.

crisbeto replied:
I'm open to this, but my thinking when I put it here was that we wanted as much host-directive-related logic inside the feature as possible.

pkozlowski-opensource reviewed (approved):
LGTM.

Would love to see cleanup I've [hinted at](https://github.com/angular/angular/pull/47476/files#r974178939) in a follow-up PR.

pkozlowski-opensource commented:
@crisbeto looks like G3 is red, not sure if this is flake or not...

crisbeto commented:
I pushed a fixup that aims to address the following issues:
* The matches array being mutated from multiple places, across different files.
* `findDirectiveDefMatches` doing both directive matching and publishing to DI.
* The host directive feature having to deal with publishing directives to DI.

AndrewKushnir commented on packages/core/src/render3/features/host_directives_feature.ts:
```diff
@@ -61,18 +61,17 @@ function findHostDirectiveDefs(
     matches: DirectiveDef<unknown>[], def: DirectiveDef<unknown>, tView: TView, lView: LView,
     tNode: TElementNode|TContainerNode|TElementContainerNode): void {
   if (def.hostDirectives !== null) {
```
Very small nit: for better minification, we can create a new const like this:
```
const hostDirectives = def.hostDirectives;
// or `const {hostDirectives} = def;` if it doesn't require downleveling...
```
so the minified code would look like this:
```
  var h = def.hostDirectives;
  if (h !== null) {
    for (let i = h.length - 1; i > -1; i--) {
      const c1 = h[i];
```
vs current code (the `hostDirectives` is retained):
```
  if (d.hostDirectives !== null) {
    for (let i = d.hostDirectives.length - 1; i > -1; i--) {
      const c = d.hostDirectives[i];
```

crisbeto replied:
I'll leave it as it is for now. In one of the later commits when we need more information, the object will get destructured in a similar way.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1291,6 +1294,8 @@ function findDirectiveDefMatches(
           matches.push(def);
         }
+
+        def.findHostDirectiveDefs?.(matches, def, tView, viewData, tNode);
```
nit: can we also add a sentence to the function description to reflect that the host directives are also appended to the list?

AndrewKushnir reviewed (approved):
FYI, just added a couple minor nits. Feel free to proceed without re-reviews from me if you decide to address the feedback (the changes would be tiny and straightforward).

crisbeto committed:
refactor(core): expose host directives to their host through DI

Exposes the host directives to the host and its descendants through DI. This can be useful, because it allows the host to further configure the host directives.

crisbeto committed:
fixup! refactor(core): expose host directives to their host through DI

crisbeto commented on packages/core/src/render3/context_discovery.ts:
```diff
@@ -295,20 +295,18 @@ function findViaDirective(lView: LView, directiveInstance: {}): number {
  */
-export function getDirectivesAtNodeIndex(
-    nodeIndex: number, lView: LView, includeComponents: boolean): any[]|null {
+export function getDirectivesAtNodeIndex(nodeIndex: number, lView: LView): any[]|null {
```
This change isn't strictly necessary, I just noticed that we weren't using this parameter anywhere.

crisbeto commented:
These changes should be good to go. Summary of the latest changes:
* I've reworked the `HostDirectivesFeature` so that it's pushing to the matches array, rather than using `unshift`.
* The `diPublicInInjector` call has been moved out of `findDirectiveDefMatches` which also allows us to remove it from the `HostDirectivesFeature`.
* I've incorporated the changes from #47490 to fix some cases in the debugging utilities where components weren't being identified correctly when they have host directives.
* I've added more unit tests to cover the lifecycle hook execution order, `ng.*` utilities and `DebugNode`.

pkozlowski-opensource commented on packages/core/test/acceptance/host_directives_spec.ts:
```diff
@@ -288,73 +291,166 @@ describe('host directives', () => {
-    ]);
+    // Note: lifecycle hook order is different when components and
+    // directives are mixed so this test aims to cover it.
+    it('should invoke lifecycle hooks in the correct order when host directives are applied to a component',
```
nit: could you replace "correct order" with something that actually describes the desired order?

something like "should invoke host directives lifecycle hooks before its host component hooks" or sth along those lines.

pkozlowski-opensource commented on packages/core/test/acceptance/host_directives_spec.ts:
```diff
@@ -288,73 +291,166 @@ describe('host directives', () => {
-      'Dir - ngAfterViewChecked',
-    ]);
+    // Note: lifecycle hook order is different when components and
+    // directives are mixed so this test aims to cover it.
```
nit: could you expand on this comment? If there is a difference I would expect to have a comment that describes 2 scenarios and highlights differences between them.

pkozlowski-opensource commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1283,15 +1286,30 @@ function findDirectiveDefMatches(
+
+          // Components are inserted at the front of the matches array so that their lifecycle
+          // hooks run before any directive lifecycle hooks. This appears to be for ViewEngine
+          // compatibility. This logic doesn't make sense with host directives, because it
```
nit: could we try to rephrase "doesn't make sense" and focus on describing how the logic should work with and without host directives?

pkozlowski-opensource reviewed (approved):
LGTM overall. Left a couple of nit comments.

I'm still not entirely sure if we could improve the logic of collecting to get rid of unshifting and I can't think of anything substantially better atm. I will try to poke around other approches tomorrow but it shouldn't block this PR.

crisbeto committed:
fixup! refactor(core): expose host directives to their host through DI

crisbeto commented:
Pushed a fixup to address the latest set of nits.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1283,15 +1286,34 @@ function findDirectiveDefMatches(
+            const hostDirectiveMatches: DirectiveDef<unknown>[] = [];
+            // Append any host directives to the matches first.
+            def.findHostDirectiveDefs(hostDirectiveMatches, def, tView, lView, tNode);
+            hostDirectiveMatches.push(def);
+            matches.unshift(...hostDirectiveMatches);
+            markAsComponentHost(tView, tNode, hostDirectiveMatches.length - 1);
```
nit, for readability (add `def` explicitly to the `matches` array, call out the `componentOffset` explicitly as well):
```suggestion
            // Add all host directives declared on this component, followed by the component itself.
            matches.unshift(...hostDirectiveMatches, def);
            // Component offset starting from the beginning of the host directives array.
            const componentOffset = hostDirectiveMatches.length;
            markAsComponentHost(tView, tNode, componentOffset);
```

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1283,15 +1286,34 @@ function findDirectiveDefMatches(
+            matches.unshift(...hostDirectiveMatches);
+            markAsComponentHost(tView, tNode, hostDirectiveMatches.length - 1);
+          } else {
+            matches.unshift(def);
+            markAsComponentHost(tView, tNode, 0);
```
```suggestion
            // No host directives on this component, just add component def to the matches.
            matches.unshift(def);
            markAsComponentHost(tView, tNode, 0);
```

AndrewKushnir commented on packages/core/test/acceptance/host_directives_spec.ts:
```diff
@@ -288,73 +291,166 @@ describe('host directives', () => {
-  it('should invoke lifecycle hooks from the host directives', () => {
-    const logs: string[] = [];
+  // TODO(crisbeto): add a test for `ngOnChanges`.
+  describe('lifecycle hooks', () => {
```
I think it'd be great to add some tests for a case when a Component has host directives that bring other host directives, so we can better cover the matching logic and component offset (if we don't have those tests already). Something like:
```
         @Directive({
           standalone: true,
           selector: '[host-dir-c]',
         })
         class HostDirC {
         }

         @Directive({
           standalone: true,
           selector: '[host-dir-b]',
           hostDirectives: [HostDirC],
         })
         class HostDirB {
         }

         @Directive({
           standalone: true,
           selector: '[host-dir-a]',
           hostDirectives: [HostDirB],
         })
         class HostDirA {
         }

         @Component({
           selector: 'comp',
           template: '',
           hostDirectives: [HostDirA],
         } as HostDirectiveAny)
         class Comp {
         }
```

AndrewKushnir reviewed (approved):
LGTM, just left a few minor comments. Comments are **not** critical and can be addressed in a followup PRs if needed. If you decide to address it in this PR, feel free to proceed without re-review from me.

Quick note on the motivation for additional comments in the code: they may feel redundant now since we have context, but I think they'd help to navigate the code later (in a few months).

crisbeto committed:
fixup! refactor(core): expose host directives to their host through DI

crisbeto reviewed (commented):
Latest set of feedback has been addressed.

alxhub commented:
This PR was merged into the repository by commit 3fe21a67ca78b04b4c9ab68fa070726e5d60e14b.
